[DRAFT]: Feature threshold 2

buffalojoec opened the pull request.

buffalojoec committed:
init new version of 0072

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,200 @@
+  features submitted for activation for a particular epoch.
+- **Support Signal PDA:** The PDA under the Feature Gate program used to store
+  a bit mask of the staged features a node supports.
+- **Feature Tombstone PDA:** The PDA under the Feature Gate program used to
+  assign accounts to, effectively "archiving" them and removing them from the
+  Feature Gate program's owned accounts.
```
Why does this need to be a PDA? Seems like some other static program id would be sufficient.

buffalojoec replied:
I was thinking in case we needed write-access to these accounts from the program context sometime in the future. If not, or if you think we can solve that another way with the runtime, sure this can be static!

CriesofCarrots replied:
I don't think using a PDA address as the program-id will work like that; the accounts would have to still be owned by the Feature Gate program for that program to write to them.
We'll need to either use runtime overrides or deploy a new core BPF program to operate on accounts owned by something other than the Feature Gate program.

buffalojoec replied:
Ah yeah true. In that case, what is the process for reserving a static address? We would need to put it in the SDK somewhere, right? And would it just be something like `FeatureTombstone11111..`?

CriesofCarrots replied:
Yeah, we would define it wherever the Feature Gate program id is defined. I don't know if that stays in solana-sdk (probably) or moves to some core-bpf repo.

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,200 @@
+
+A node signals its support for staged features by invoking the Feature Gate
+program's `SignalSupportForStagedFeatures` instruction, which requires a  
+`[u8; 5]` bit mask of the staged features. Each element of the bit mask is
```
It's not really a "bit" mask if you're using a full byte for each item. Use just one byte instead? This would allow us up to 255 features, if we change the limit in the future. That seems like more than enough.

buffalojoec replied:
Sure, `u8` will work! Do you think the cap of 5 for now makes sense? I sort of pulled out an arbitrary number to gauge interest in the proper cap.

CriesofCarrots replied:
I'm fine with using 5 to start with, and we can see what people think. Up to you, but you could mention the fact that the cap could be raised up to 255 if we feel a lot more aggressive about activations in the future.

buffalojoec replied:
Btw, I think this would actually cap us at 8 features, ie `11111111`, since the ordering of staged features may not correlate with software versioning. Someone might signal support for 4 of 8 features and send a `10101010`.

CriesofCarrots replied:
Oh duh, sorry, that's right. Maybe just change the proposal to cap at 8, then?

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,200 @@
+- **Feature Gate program:** The Core BPF program introduced in
+  [SIMD 0089](https://github.com/solana-foundation/solana-improvement-documents/pull/89)
+  that will own all feature accounts.
+- **Staged Features PDA:** The PDA under the Feature Gate program used to track
+  features submitted for activation for a particular epoch.
+- **Support Signal PDA:** The PDA under the Feature Gate program used to store
+  a bit mask of the staged features a node supports.
```
Based on my experience with SIMD 0106, I think we'll need to define the addresses of these PDAs somewhere in this proposal. This will help clarify why the Staged Features PDA of one epoch (which validators are currently considering for support) doesn't collide with the Staged Features PDA of the next epoch (which is being actively added to as features are staged).

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,200 @@
+
+- Archive any activated feature accounts
+- Archive this epoch's Staged Features PDA
+- Delete any spam feature accounts with no state
```
I think this one needs a little more explanation. Might not be spam, just a feature that didn't get staged this epoch.

buffalojoec replied:
Good shout. Based on our convo offline about creating vs. staging, this section will need a little touching up.

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,200 @@
+validators.
+
+Core contributors will no longer bear the responsibility of ensuring the proper
+stake supports their feature activation. However, it will change the process
+by which they can override this stake requirement to push a feature through.
```
What is the process to override the stake requirement? (I don't think there is one yet?)

buffalojoec replied:
Yeah I intentionally left it out. If you want it in here we can add it, but conceptually it would probably be some additional `Feature` state, or something similar the program can write.

CriesofCarrots replied:
In the interest of honesty, I think say that under this proposal, there is no way to override the stake requirement, but that could/should be proposed separately...?

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,200 @@
+
+### Conclusion
+
+This new process provides additional safeguards for submitting features for
+activation as well as an automated stake support threshold check before
+activating them. This greatly reduces the risk of human error when enabling new
+functionality on the network, as well as minimizes the risk of cluster
+partitioning.
+
+Additionally, this system makes use of on-chain data and a client-agnostic BPF
+program to perfom its duties, further decentralizing the feature activation
+process.
+
+These enhancements will be especially useful in a world where multiple clients
+will aim to push features and seek to agree on their activation.
```
I think I would remove this section and split the content between Motivation, Impact, and Security Considerations. (Although I think you already say some of this there)

buffalojoec committed:
use static address for tombstone

buffalojoec committed:
define PDAs

buffalojoec committed:
use  for bit mask

buffalojoec committed:
clean up closing remarks

buffalojoec committed:
open feature creation and garbage collection

buffalojoec committed:
wording touch-ups

buffalojoec committed:
drop instruction layout

buffalojoec commented:
I think this is ready to go back into the main repo for discussion.

I dropped the instruction layout, since I've been typically advised to exclude code snippets, but I can add it back if need be.

I'll squash these commits, but I'm curious how best to re-introduce this SIMD? I could:
- Squash into one commit on the original branch and write a message in the original PR
- Create a new SIMD altogether

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+based on a stake-weighted support threshold, rather than manual human action.
+
+With this new process, contributors would no longer have to assess stake support
+for a feature before activation. Instead, this would be done by the runtime.
```
```suggestion
With this new process, contributors no longer have to assess stake support
for a feature before activation. Instead, the assessment is done by the runtime.
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+
+## Detailed Design
+
+The proposed new process would be comprised of the following steps:
```
```suggestion
The new process is comprised of the following steps:
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+1. **Feature Creation:** Contributors create feature accounts with the Solana
+   CLI as they do now.
+2. **Staging Features for Activation:** In some epoch 0, the multi-signature
+   authority stages created features for activation.
```
```suggestion
2. **Staging Features for Activation:** In some epoch N, a multi-signature
   authority stages for activation some or all of the features created in step 1.
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+   CLI as they do now.
+2. **Staging Features for Activation:** In some epoch 0, the multi-signature
+   authority stages created features for activation.
+3. **Signaling Support for Staged Features:** During the next epoch (epoch 1),
```
```suggestion
3. **Signaling Support for Staged Features:** During the next epoch (epoch N+1),
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+3. **Signaling Support for Staged Features:** During the next epoch (epoch 1),
+   validators signal which of the staged feature-gates they support in their
+   software.
+4. **Activation and Garbage Collection:** On the next epoch boundary, the
```
```suggestion
4. **Activation and Garbage Collection:** At the end of epoch N+1, the
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+
+The first step in the new process shall consist of the already existing method
+for creating new features using the Solana CLI (`solana feature activate`). This
+command could possibly be renamed to `solana feature create`.
```
I think I would remove references to the CLI here, as the CLI is an implementation detail. It doesn't matter how a user creates a feature account. Essentially just remove this entire paragraph.

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+The proposed new process would be comprised of the following steps:
+
+1. **Feature Creation:** Contributors create feature accounts with the Solana
+   CLI as they do now.
```
```suggestion
1. **Feature Creation:** Contributors create feature accounts as they do now.
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+for creating new features using the Solana CLI (`solana feature activate`). This
+command could possibly be renamed to `solana feature create`.
+
+This command sends a transaction containing System instructions to fund,
```
```suggestion
The first step is creation of a feature account, done by submitting a transaction containing System instructions to fund,
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+allocate, and assign the feature account to
+`Feature111111111111111111111111111111111111`. This can be performed by anyone.
+
+This step merely creates a feature, but will no longer stage it for activation.
```
I think we can remove this as well.

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+A multi-signature authority shall be created to stage features for activation.
+This multi-signature will comprise key-holders from Solana Labs and possibly
+from other validator client teams in the future. In the future, this authority
+could be replaced by validator governance.
```
```suggestion
A multi-signature authority, comprising key-holders from Solana Labs and possibly
other validator client teams in the future, will have the authority to stage created features for activation.
In the future, this authority could be replaced by validator governance.
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+could be replaced by validator governance.
+
+The multi-signature authority stages a feature for activation by invoking the
+Feature Gate program's `StageFeatureForActivation` instruction, which verifies
```
I know you explicitly dropped the instruction-layout stubs, but I think we should probably define in prose the accounts and data that the instruction expects. Maybe before "which verifies", which could become something like: "The processor for this instruction must verify..."

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+**next epoch's** Staged Features PDA.
+
+The Staged Features PDA for a given epoch stores a list of all feature IDs that
+were staged for activation during the previous epoch. To start, this list shall
```
```suggestion
were staged for activation during the previous epoch. This list shall
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+The proposed seeds for deriving the Staged Features PDA are provided below,
+where the `<epoch>` represents the epoch during which the contained feature IDs
+will be assessed for stake support and potentially activated at the end of that
+epoch.
```
```suggestion
where the `<epoch>` is the epoch during which the contained feature IDs
will be assessed for stake support and considered for activation.
```

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+supported by their software.
+
+A node signals its support for staged features by invoking the Feature Gate
+program's `SignalSupportForStagedFeatures` instruction, which requires a  
+`u8` bit mask of the staged features.
```
Same comment about expected accounts and data, ie. make more clear that the "`u8` bit mask of the staged features" is expected data.

CriesofCarrots commented on proposals/feature-threshold-automation.md:
````diff
@@ -0,0 +1,197 @@
+"support_signal" + <vote_address>
+```
+
+Nodes shall send this instruction at some arbitrary point during the epoch at
````
```suggestion
Nodes should send a transaction containing this instruction at some arbitrary point during the epoch at
```

buffalojoec replied:
You don't really like "shall" do you? 😂

CriesofCarrots replied:
Haha, I am realizing that about myself ;)
However, in this particular case, I had just read rfc2119 (due to Trent's prompting on my SIMD), and I think "should" is actually the correct word.

CriesofCarrots commented on proposals/feature-threshold-automation.md:
```diff
@@ -0,0 +1,197 @@
+
+The runtime "archives" an account by assigning it to the Feature Tombstone.
+
+Created features that were not staged for activation or did not meet the
+required stake support will not be garbage collected.
```
This seems to contradict the 3rd bullet point, because I think features not staged for activation will be feature accounts with no state owned by the Feature Gate program, right?

buffalojoec replied:
Hm, maybe I should be more clear on the third bullet and here. The only reliable way to isolate spam would be to check against the feature set listed in the SDK. I think there's no reason a real feature ID would exist outside of that list, right? If so, then this gets a little tricky to garbage collect.

CriesofCarrots replied:
Perhaps I don't understand your intent. I thought you were considering as spam anything that did not get staged during the epoch.
I don't think we should depend on the feature set listed in the SDK for any kind of assessment of the accounts, since we're trying to divorce the activation process from that list. I think we want the behavior here to be the same for all nodes, even if a feature isn't present in the client software (yet).

buffalojoec replied:
Makes sense. Sounds like the best way to identify garbage features is to require the minimum deposit you've suggested. The question is... how much? 👀

buffalojoec replied:
Without the deposit, we can't reasonably identify a feature account as spam, because legitimate features created by anyone will have the same size and rent lamports. No state.

CriesofCarrots replied:
Right. We either need to leave all of these empty feature accounts or clean them all up at the epoch boundary. Since it won't grief the staging process and is consistent with the current state of things, I'm leaning toward the first option.

buffalojoec replied:
Wait, it sounds like you're saying don't even bother with garbage collection?

CriesofCarrots replied:
We should definitely consider not bothering.
Right now, anyone can create a keypair and fund and assign the account at that pubkey to some random program, as long as the amount they fund is >= the rent-exempt minimum for an empty account. So defunding such feature program accounts doesn't really do much to prevent spam in the grand scheme of things. The only reason to do it is if we need to scan all feature accounts for something, or if having spam feature accounts prevents some action we want to take on legitimate accounts. Afaict, our current design avoids both of these things.

CriesofCarrots replied:
That said, we could always implement garbage collection of such accounts later. And since the accounts are owned by the feature program, we could actually support a manual cleanup with just an upgrade to the Feature Gate program.

buffalojoec replied:
Ok yeah that sounds good to me. If we don't have a direct need to garbage collect, let's take it out. It reduces complexity and compute.

I agree, in our proposal here and the current runtime implementation we don't load the Feature Gate accounts en masse. I was curious if we did this in the runtime, but this appears to be the feature set's entrypoint for a `Bank` instance:

https://github.com/solana-labs/solana/blob/daa2449ad4b4dc0cc83a4be1968653815e2e4a8c/runtime/src/bank.rs#L1024

https://github.com/solana-labs/solana/blob/daa2449ad4b4dc0cc83a4be1968653815e2e4a8c/sdk/src/feature_set.rs#L1021

CriesofCarrots replied:
To be clear, I do still think we should "archive" activated features and the staged features PDAs.

buffalojoec replied:
Agreed, just not the potentially "spam" feature accounts.

CriesofCarrots reviewed (commented):
I did one more pass, more nit-picky about language and such.

> I think this is ready to go back into the main repo for discussion.

I agree!

> I'll squash these commits, but I'm curious how best to re-introduce this SIMD? I could:
> -   Squash into one commit on the original branch and write a message in the original PR
 > -  Create a new SIMD altogether

Hmm, I'm torn. It is the same proposal, just rewritten. But I'm leaning toward new SIMD, since it will simplify the discussion and we can still link back to the original. Maybe ask Jacob what he thinks is most in line with the spirit of the SIMD process?

buffalojoec committed:
review feedback 1-31

buffalojoec committed:
remove spam garbage collection

buffalojoec committed:
small corrections

buffalojoec commented:
Alrighty! I've addressed everything. Thanks for the reviews and helping me craft this new version! Let's send it in.

> Hmm, I'm torn. It is the same proposal, just rewritten. But I'm leaning toward new SIMD, since it will simplify the discussion and we can still link back to the original. Maybe ask Jacob what he thinks is most in line with the spirit of the SIMD process?

I spoke with Jacob and he advised using the same proposal, so I'll try to clean up the discussions and throw a flare up in there when I push the squashed commit.

One last note: Should we swap mentions of "Solana Labs maintainers" with "Anza maintainers"?

CriesofCarrots commented:
> One last note: Should we swap mentions of "Solana Labs maintainers" with "Anza maintainers"?

Yeah, I was wondering that too. I think leave it as Labs until the actual move to the new monorepo has happened? I realize that's supposed to be soon, and just adds some churn, so I don't feel strongly about that.
